crates/re_types_builder/src/codegen/cpp.rs

```diff
@@ -167,3 +191,609 @@ fn pragma_once() -> TokenStream {
     }
 }
+
+struct QuotedObject {
+    hpp: TokenStream,
+    cpp: TokenStream,
+}
+
+impl QuotedObject {
+    pub fn new(objects: &Objects, obj: &crate::Object) -> Self {
+        match obj.specifics {
+            crate::ObjectSpecifics::Struct => Self::from_struct(objects, obj),
+            crate::ObjectSpecifics::Union { .. } => Self::from_union(objects, obj),
+        }
+    }
+
+    fn from_struct(_objects: &Objects, obj: &crate::Object) -> QuotedObject {
+        let namespace_ident = format_ident!("{}", obj.kind.plural_snake_case());
+        let pascal_case_name = &obj.name;
+        let pascal_case_ident = format_ident!("{pascal_case_name}");
+        let quoted_docs = quote_docstrings(&obj.docs);
+
+        let mut hpp_includes = Includes::default();
+
+        let field_declarations = obj
+            .fields
+            .iter()
+            .map(|obj_field| {
+                let declaration = quote_declaration_with_docstring(
+                    &mut hpp_includes,
+                    obj_field,
+                    &format_ident!("{}", obj_field.name),
+                    false,
+                )
+                .0;
+                quote! {
+                    #NEWLINE_TOKEN
+                    #declaration
+                }
+            })
+            .collect_vec();
+
+        let constructor = if obj.fields.len() == 1 {
+            // Single-field struct - it is a newtype wrapper.
+            // Create a implicit constructor from its own field-type.
+            let obj_field = &obj.fields[0];
+            if let Type::Array { .. } = &obj_field.typ {
+                // TODO(emilk): implicit constructor for arrays
+                quote! {}
+            } else {
+                hpp_includes.system.insert("utility".to_owned()); // std::move
+
+                let field_ident = format_ident!("{}", obj_field.name);
+                let parameter_declaration =
+                    quote_declaration(&mut hpp_includes, obj_field, &field_ident, false).0;
+                quote! {
+                    #pascal_case_ident(#parameter_declaration) : #field_ident(std::move(#field_ident)) {}
+                }
+            }
+        } else {
+            quote! {}
+        };
+
+        let hpp = quote! {
+            #hpp_includes
+
+            namespace rr {
+                namespace #namespace_ident {
+                    #quoted_docs
+                    struct #pascal_case_ident {
+                        #(#field_declarations;)*
+
+                        #constructor
+                    };
+                }
+            }
+        };
+
+        let cpp = quote! {};
+
+        Self { hpp, cpp }
+    }
+
+    fn from_union(objects: &Objects, obj: &crate::Object) -> QuotedObject {
+        // We implement sum-types as tagged unions;
+        // Putting non-POD types in a union requires C++11.
+        //
+        // enum class Rotation3DTag {
+        //     NONE = 0, // Makes it possible to implement move semantics
```

Comment:
Exactly. Move semantics in C++ requires one can leave an object in an "empty" state so that the destructor can safely run. For something like `std::string` the empty state is just the empty string, but for something like our union we really do need an extra variant.

The good news is this also makes it easier to implement https://github.com/rerun-io/rerun/issues/2684